ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java

```diff
@@ -349,34 +356,60 @@ private void onErrorResponse(ReplicaId replicaId, RouterException exception, boo
     }
   }
 
+  /**
+   * Check if we should replicate the blob and retry the operation.
+   * @return true if need to do the retry.
+   */
+  private boolean shouldReplicateBlobAndRetry() {
+    // the conditions to trigger ReplicateBlob and retry the original operation:
+    // 1. the feature is enabled.
+    // 2. and at least one replica returned NOT_FOUND status.
+    // 3. and at least one replica returned successful status.
+    return (routerConfig.routerRepairWithReplicateBlobEnabled && operationTracker.hasNotFound()
+        && operationTracker.getSuccessCount() > 0);
```

Comment:
I don't think this is the right condition. If you have one success and one one delete, this is probably not retry-able anymore. I think the right condition should be that we don't an error code that has precedence over AmbryUnavailable. Something like
```
RouterErrorCode errorCode = ((RouterException)operationException.get()).getErrorCode();
return routerConfig.ReplicateBlobEnabled && operationTracker.getSuccessCount() > 0 && getPrecedence(errorCode) >= getPrecedence(RouterErrorCode.AmbryUnavailable);
```

Reply:
You are right. Thanks for catching this.

Added the error code precedence condition. (Sorry I didn't get it from your first comment about it.)
I think we should still keep hasNotFound condition because ReplicateBlob only helps if there is some replica returned NOT_FOUND. If 8 replicas are unavailable, won't help either.